### PR TITLE
Description & tagging on dashboards

### PR DESCRIPTION
**Pull request by paolodamico**
## Changes

This PR adds more stuff from #3344 & #3551. Please view original issues for motivation and context.
- Allows adding descriptions to dashboards to better provide context on each dashboard.
    
    
- Allows tagging dashboards for easier organization (includes autocomplete, duplicates prevention, loading states, ...).
    image
- Tags are shown on dashboard list too and can be filtered on.
    
    
- Descriptions are shown when publicly sharing a dashboard.
     
- Empty state when you don't have any tags registered.
    
- Adjusts relevant instrumentation accordingly.

## Checklist

- [X] All querysets/queries filter by Organization, by Team, and by User
- [X] Django backend tests
- [X] Jest frontend tests
- [X] Cypress end-to-end tests

### CODE REVIEW

**Comment by paolodamico** [comment]
@jamesefhawkins this seems like a feature geared towards collaboration, should we make it a premium feature? I'm worried it's just a tiny thing that will cause more complexity (UX & code) than what is worth. Thoughts?

**Comment by paolodamico** [comment]
UX question for @corywatilo, opted for allowing direct changes to tags without entering edit mode, wdyt?

File: frontend/src/scenes/dashboard/dashboardsLogic.ts [hunk 1]
```diff
@@ -25,6 +26,18 @@ export const dashboardsLogic = kea<dashboardsLogicType>({
                     .filter((d) => !d.deleted)
                     .sort((a, b) => (a.name ?? 'Untitled').localeCompare(b.name ?? 'Untitled')),
         ],
+        dashboardTags: [
+            () => [dashboardsModel.selectors.dashboards],
+            (dashboards: DashboardType[]): string[] =>
+                dashboards.length
+                    ? uniqueBy(
+                          dashboards
+                              .map((dashboard) => dashboard.tags.map((tag) => tag))
```
Comment (macobo) [inline frontend/src/scenes/dashboard/dashboardsLogic.ts:35]: This chain can be simplified to:

```
dashboard.flatMap(({tags} => tags)
```

I'd also consider adding a default argument to `uniqueBy`

File: frontend/src/scenes/dashboard/dashboardsLogic.ts [hunk 2]
```diff
@@ -25,6 +26,18 @@ export const dashboardsLogic = kea<dashboardsLogicType>({
                     .filter((d) => !d.deleted)
                     .sort((a, b) => (a.name ?? 'Untitled').localeCompare(b.name ?? 'Untitled')),
         ],
+        dashboardTags: [
+            () => [dashboardsModel.selectors.dashboards],
+            (dashboards: DashboardType[]): string[] =>
+                dashboards.length
```
Comment (macobo) [inline frontend/src/scenes/dashboard/dashboardsLogic.ts:32]: No need for ternary

File: frontend/src/scenes/dashboard/dashboardLogic.js [hunk 3]
```diff
@@ -20,7 +20,7 @@ export const dashboardLogic = kea({
```
Comment (macobo) [inline frontend/src/scenes/dashboard/dashboardLogic.js:20]: Maybe worth converting to ts as you're editing this?

**Review by macobo — COMMENTED** [review]
Added some minor nits, did not go through the whole flow though

**Comment by jamesefhawkins** [comment]
> @jamesefhawkins this seems like a feature geared towards collaboration, should we make it a premium feature? I'm worried it's just a tiny thing that will cause more complexity (UX & code) than what is worth. Thoughts?

I definitely think this has to be paid so that we can keep giving a ton of functionality to free users whilst building a real business at the same time. Collaboration-related features like this definitely aren't needed for 1 person, but for a team of more than a couple (thus a company, with money involved) are really important.

I can see how this is annoying code-wise, but this will happen repeatedly, so we need to find a way to do easily if we can?

File: frontend/src/scenes/dashboard/dashboardsLogic.ts (hunk 1 above)
Comment (paolodamico) [reply frontend/src/scenes/dashboard/dashboardsLogic.ts:35]: Magic, thanks for the tip!

File: frontend/src/scenes/dashboard/dashboardLogic.js (hunk 3 above)
Comment (paolodamico) [reply frontend/src/scenes/dashboard/dashboardLogic.js:20]: Will leave this for a separate PR, there are too many changes required (60+)

**Comment by paolodamico** [comment]
Makes sense @jamesefhawkins, will make it a paid feature, we do have a way to handle this that is fairly simple to implement, it's just something we'll have to mind when we make changes to this feature.

**Pull request merged**